operator/internal/status/status.go

```diff
@@ -2,69 +2,81 @@ package status
-func Refresh(ctx context.Context, k k8s.Client, req ctrl.Request) error {
-	if err := SetComponentsStatus(ctx, k, req); err != nil {
-		return err
-	}
-
-	var s lokiv1.LokiStack
-	if err := k.Get(ctx, req.NamespacedName, &s); err != nil {
+func Refresh(ctx context.Context, k k8s.Client, req ctrl.Request, now time.Time) error {
+	var stack lokiv1.LokiStack
+	if err := k.Get(ctx, req.NamespacedName, &stack); err != nil {
 		if apierrors.IsNotFound(err) {
 			return nil
 		}
 		return kverrors.Wrap(err, "failed to lookup lokistack", "name", req.NamespacedName)
 	}
 
-	cs := s.Status.Components
+	cs, err := generateComponentStatus(ctx, k, &stack)
+	if err != nil {
+		return err
+	}
+
+	condition := generateCondition(cs)
 
-	// Check for failed pods first
-	failed := len(cs.Compactor[corev1.PodFailed]) +
-		len(cs.Distributor[corev1.PodFailed]) +
-		len(cs.Ingester[corev1.PodFailed]) +
-		len(cs.Querier[corev1.PodFailed]) +
-		len(cs.QueryFrontend[corev1.PodFailed]) +
-		len(cs.Gateway[corev1.PodFailed]) +
-		len(cs.IndexGateway[corev1.PodFailed]) +
-		len(cs.Ruler[corev1.PodFailed])
+	condition.LastTransitionTime = metav1.NewTime(now)
+	condition.Status = metav1.ConditionTrue
 
-	unknown := len(cs.Compactor[corev1.PodUnknown]) +
-		len(cs.Distributor[corev1.PodUnknown]) +
-		len(cs.Ingester[corev1.PodUnknown]) +
-		len(cs.Querier[corev1.PodUnknown]) +
-		len(cs.QueryFrontend[corev1.PodUnknown]) +
-		len(cs.Gateway[corev1.PodUnknown]) +
-		len(cs.IndexGateway[corev1.PodUnknown]) +
-		len(cs.Ruler[corev1.PodUnknown])
+	statusUpdater := func(stack *lokiv1.LokiStack) {
+		stack.Status.Components = *cs
 
-	if failed != 0 || unknown != 0 {
-		return SetFailedCondition(ctx, k, req)
-	}
+		index := -1
+		for i := range stack.Status.Conditions {
+			// Reset all other conditions first
+			stack.Status.Conditions[i].Status = metav1.ConditionFalse
+			stack.Status.Conditions[i].LastTransitionTime = metav1.NewTime(now)
 
-	// Check for pending pods
-	pending := len(cs.Compactor[corev1.PodPending]) +
-		len(cs.Distributor[corev1.PodPending]) +
-		len(cs.Ingester[corev1.PodPending]) +
-		len(cs.Querier[corev1.PodPending]) +
-		len(cs.QueryFrontend[corev1.PodPending]) +
-		len(cs.Gateway[corev1.PodPending]) +
-		len(cs.IndexGateway[corev1.PodPending]) +
-		len(cs.Ruler[corev1.PodPending])
+			// Locate existing pending condition if any
+			if stack.Status.Conditions[i].Type == condition.Type {
+				index = i
+			}
+		}
 
-	if pending != 0 {
-		return SetPendingCondition(ctx, k, req)
+		if index == -1 {
+			stack.Status.Conditions = append(stack.Status.Conditions, condition)
+		} else {
+			stack.Status.Conditions[index] = condition
+		}
+	}
+
+	statusUpdater(&stack)
+	err = k.Status().Update(ctx, &stack)
+	switch {
+	case err == nil:
+		return nil
+	case apierrors.IsConflict(err):
+		// break into retry-logic below on conflict
+		break
+	case err != nil:
+		// return non-conflict errors
+		return err
```

Comment:
Maybe I've missed something, but do we even need this when we just call `retry.RetryOnConflict` anyway?

Reply:
The flow of this function is a bit unintuitive and that's something that still bothers me. The `retry.RetryOnConflict` will only be reached in the second case because of the `break`, the other two cases return from this function directly.

The upshot of this is that the retry-logic and with that the second `Get` call are only triggered if there was a conflict on the first try.

Reply:
Oh, sorry. The difference a single word makes. 😆 

I meant: Maybe I've missed something, but do we even need this when we **can** just call `retry.RetryOnConflict` anyway?